### ISSUE TITLE
FN `attribute-defined-outside-init` with `setattr()`

### ISSUE THREAD

**Issue by jacobtylerwalls** [pylint-dev/pylint#9798]
### Bug description

```python
# pylint: disable=missing-module-docstring, too-few-public-methods
class NoAttributesDefinedInInit:
    """Both assignments should raise."""
    def later(self):
        """Only the first assignment raises."""
        self.a = "a"
        setattr(self, "b", "b")
```

### Command used

```shell
pylint a.py
```

### Pylint output

```python
************* Module a
pylint/a.py:6:8: W0201: Attribute 'a' defined outside __init__ (attribute-defined-outside-init)

------------------------------------------------------------------
Your code has been rated at 7.50/10 (previous run: 7.50/10, +0.00)
```

### Expected behavior

Two messages. You shouldn't be able to get around the lint warning by using `setattr()` (unless the attribute was already defined in `__init__()`.

### Pylint version

```shell
pylint 3.3.0-dev0
astroid 3.3.0-dev0
Python 3.12.1
```
